mmap-alloc/src/lib.rs

```diff
@@ -759,6 +842,43 @@ mod tests {
     }
 
     #[test]
+    fn test_small_realloc() {
+        unsafe {
+            let mut allocator = MapAlloc::default();
+            let small = Layout::array::<u8>(1).unwrap();
+            let medium = Layout::array::<u8>(2048).unwrap();
+            let big = Layout::array::<u8>(4096).unwrap();
+            let ptr = <MapAlloc as Alloc>::alloc(&mut allocator, medium.clone()).unwrap();
+            allocator.shrink_in_place(ptr, medium.clone(), small.clone()).unwrap();
+            allocator.grow_in_place(ptr, small.clone(), big.clone()).unwrap();
+            let old_ptr = ptr;
+            let ptr = allocator.realloc(ptr, big.clone(), small.clone()).unwrap();
+            assert_eq!(old_ptr, ptr);
+            <MapAlloc as Alloc>::dealloc(&mut allocator, ptr, small.clone());
+        }
+    }
+
+    #[cfg(target_os = "linux")]
+    #[test]
+    fn test_large_realloc() {
+        unsafe {
+            let mut allocator = MapAlloc::default();
+            let small = Layout::array::<u8>(1).unwrap();
+            let medium = Layout::array::<u8>(allocator.pagesize * 8).unwrap();
+            let big = Layout::array::<u8>(allocator.pagesize * 16).unwrap();
+            let ptr = <MapAlloc as Alloc>::alloc(&mut allocator, big.clone()).unwrap();
+            allocator.shrink_in_place(ptr, big.clone(), small.clone()).unwrap();
+            allocator.grow_in_place(ptr, small.clone(), medium.clone()).unwrap();
+            let old_ptr = ptr;
+            let ptr = allocator.realloc(ptr, medium.clone(), small.clone()).unwrap();
+            assert_eq!(old_ptr, ptr);
+            let ptr = allocator.realloc(ptr, small.clone(), big.clone()).unwrap();
+            assert_eq!(old_ptr, ptr);
+            <MapAlloc as Alloc>::dealloc(&mut allocator, ptr, big.clone());
+        }
+    }
```

Comment:
In both of these tests, can you add calls to `test_valid_map_address`, `test_zero_filled`, `test_write`, and `test_write_read` as appropriate to ensure that the returned allocation really behaves as expected (we can access it w/o segfault, the memory beyond the old end of the allocation is zero-filled, etc)?

Also, it'd be good to test to make sure that when an allocation is moved (if it can't be expanded in place), the contents of the old allocation are copied to the new one. I think what could work here would be something like:

```rust
let two_pages = map(2 * pagesize(), PROT_READ_WRITE, true, None).unwrap();
let mut alloc = MapAlloc::default();

// fill the first page with unique data that we can check later

// Treating the first page as its own allocation, grow that allocation to two pages.
// Since we know there's a second mapped page right after it, remap will be unable
// to simply grow the mapping in place, and will be forced to move it.
let new = alloc.realloc(two_pages, Layout::array::<u8>(pagesize()).unwrap(), Layout::array::<u8>(2 * pagesize()).unwrap()).unwrap();

// verify that both pages can be accessed w/o segfault, that the second page is zero-filled, and that the first page's contents have been properly copied from the original allocation
```